internal/resolver/config_selector.go

```diff
@@ -0,0 +1,72 @@
+/*
+ *
+ * Copyright 2020 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package resolver provides internal resolver-related functionality.
+package resolver
+
+import (
+	"context"
+	"sync"
+
+	"google.golang.org/grpc/internal/serviceconfig"
+	"google.golang.org/grpc/resolver"
+)
+
+// ConfigSelector controls what configuration to use for every RPC.
+type ConfigSelector interface {
+	// Selects the configuration for the RPC.
+	SelectConfig(RPCInfo) *RPCConfig
+}
+
+// RPCInfo contains RPC information needed by a ConfigSelector.
+type RPCInfo struct {
+	Context context.Context // Contains headers, application timeout
```

Comment:
Is it possible to move this context out of here and instead make it the first parameter to `SelectConfig`?

Reply:
I really don't want to do that, as the config selector is not supposed to be blocking and having a `Context` parameter makes it look like it is fine to block.  `Context` is only passed here for the data it contains (metadata & RPC deadline primarily).  This is now awkward because gRPC-Go's API used context inappropriately, but it's better to pass it in a struct than as the first parameter to avoid assumptions about what it means IMO.

Reply:
Agree with the blocking aspect. But my concern is the same as that in the other comment, that this context might become a dumping ground for stuff. And for someone reading this part of the code, it is not super clear what exact things are passed in this context. They would have to go look at the usages of this context field.

Reply:
This is the user's context for the RPC.  So it has whatever the user put in there, of which gRPC only cares about deadline and metadata (unless I'm forgetting something).  I wouldn't advocate adding anything else there; we'd use `CallOption`s instead (what should have been used for metadata and deadline anyway so we wouldn't need to pass this).  If we pull out the metadata and pass it instead, then we would be wasting resources (allocations & copies) for something that isn't used by the default config selector (service config).

Also, this is an internal API, so I think we get away with taking some liberties here.

Reply:
Maybe add a comment to explain the rationale. Thanks.

Reply:
Done